src/libraries/ZkAddress.sol

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: CC0-1.0
+
+pragma solidity 0.8.15;
+
+import "@base58-solidity/Base58.sol";
+
+library ZkAddress {
+    error InvalidZkAddressLength();
+    error InvalidZkAddressChecksum();
+
+    struct ZkAddress {
+        bytes10 diversifier;
+        bytes32 pk;
+    }
+
+    function parseZkAddress(bytes memory _rawZkAddress, uint24 _poolId) external pure returns (ZkAddress memory) {
+        uint256 len = _len(_rawZkAddress);
+        if (len > 64 || (len < 46 && len != 42)) {
+            revert InvalidZkAddressLength();
+        }
+
+        // _zkAddress == abi.encodePacked(bytes10(diversifier), bytes32(pk))
+        if (len == 42) {
+            return ZkAddress(bytes10(_load(_rawZkAddress, 0)), _load(_rawZkAddress, 10));
+        }
+
+        // _zkAddress == abi.encode(bytes10(diversifier), bytes32(pk))
+        if (len == 64) {
+            return abi.decode(_rawZkAddress, (ZkAddress));
+        }
```

Comment:
Are you sure these cases are safe? I suppose we shouldn't accept direct deposit without zkAddress validation. Otherwise the user can lose his money by sending on incorrect address.

Reply:
I think it's still might be beneficial to support multiple formats. Though it makes sense to recommend checksummed addresses for most of the use-cases, that would be highlighted in docs. But if the address correctness was validated on UI side, then I don't see a big reason to repeat it's validation in the contract.

Anyway, we cannot prevent loss of funds in all cases, in the same way we can't prevent usage of public 0x address which do not belong to anyone.

Reply:
But what if the endianness was corrupted in core library due to any error?

Reply:
But we can calculate checksum for already corrupted address by the same logic, can't we?